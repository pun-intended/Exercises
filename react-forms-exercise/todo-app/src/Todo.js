 import "./Todo.css"

 function Todo({task, id, remove}){
    function handleSubmit(evt){
        evt.preventDefault();
        remove(id);
    }
    return(
        <div className="Todo" id={id}>
            <form onSubmit={handleSubmit}>
                <p className="Todo-task">{task}</p><button className="Todo-button">X</button>
            </form>
        </div>
    )
 }

 export default Todo