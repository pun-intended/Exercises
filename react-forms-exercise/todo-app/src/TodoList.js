import React, { useState } from "react"
import Todo from "./Todo.js"
import NewTodoForm from "./NewTodoForm.js"
import {v4 as uuid} from 'uuid'

function TodoList(){
    const [todos, setTodos] = useState([])

    function addTodo(data){
        const newItem = {
            task: data.task,
            id: uuid()
        }
        setTodos([...todos, newItem])
    }

    function removeTodo(id){
        const filteredTodos = todos.filter((t) => t.id !== id)
        setTodos([...filteredTodos])
    }

    return (
        <div className="TodoList">
            <NewTodoForm addTodo={addTodo} />
            {todos.map(t => {return(
                <Todo task={t.task} id={t.id} remove={removeTodo} key={t.id}/>
            )})}
        </div>
    )
}

export default TodoList


