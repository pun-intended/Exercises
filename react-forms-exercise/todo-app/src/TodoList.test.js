import { render, screen } from '@testing-library/react';
import TodoList from './TodoList';
import Todo from './Todo';

function addTodo(list, task="new task"){
    const taskInput = list.getByLabeltext("Task:")
    fireEvent.change(taskInput, {target: {value: task}});
    const btn = list.getByText("Add")
    fireEvent.click(btn)
}

// Smoke test
render(<TodoList />)

// Snapshot test
const { asFragment } = render(<TodoList />)
expect(asFragment()).toMatchSnapshot()

// Can add item
test("Can add todos", () => {
    const list = render(<TodoList />)
    addTodo(list)

    // Expect task to be added and input cleared
    expect(list.getByLabelText("Task:")).toHaveValue("")
    expect(list.getByText("new task")).toBeInTheDocument()
    expect(list.getByText("X")).toBeInTheDocument()
})
// Can delete item
test("Can delete todos", () => {
    const list = render(<TodoList />)
    addTodo(list)

    fireEvent.click(list.getByText("X"))
    expect(list.queryByText("new task")).not.toBeInTheDocument()
})