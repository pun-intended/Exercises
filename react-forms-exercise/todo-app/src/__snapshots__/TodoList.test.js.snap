// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <div>
      <form>
        <label
          for="task"
        >
          Task
        </label>
        <input
          name="task"
          value=""
        />
        <button>
          Add
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
